### PR TITLE
CellUnion: parity with upstream S2 (closes #76)

### LINKED ISSUE #76: Finish CellUnion
Finish the code in `s2/cellunion.rs` to make it equivalent to the C++ and Go versions.

We have some commented-out Go code in our repo. However, that copy was made a long time ago, so it might have digressed from C++ and Go. Might be good to double-check (at least the test cases), especially against the C++ version which is the ground truth.

https://github.com/google/s2geometry/blob/master/src/s2/s2cell_union.h
https://github.com/google/s2geometry/blob/master/src/s2/s2cell_union.cc
https://github.com/google/s2geometry/blob/master/src/s2/s2cell_union_test.cc

https://github.com/golang/geo/blob/master/s2/cellunion.go
https://github.com/golang/geo/blob/master/s2/cellunion_test.go

### PR DESCRIPTION
**Pull request by Louisvranderick**
## Summary

Brings `CellUnion` in line with the geometric API and behavior of [google/s2geometry](https://github.com/google/s2geometry) (`s2cell_union`) and [golang/geo](https://github.com/golang/geo) (`cellunion`), as requested in #76.

## Implemented

- **Validity / shape**: `is_valid`, `is_normalized`
- **Set-style ops on normalized unions**: `merge`, `union`, `intersection`, `difference`, `intersection_with_cell_id`
- **Relations**: `contains_cell_union`, `intersects_cell_union`, `contains_point`
- **Expansion**: `expand_at_level`, `expand_by_radius` (using `MIN_WIDTHMETRIC`)
- **Areas**: `average_area`, `approx_area`, `exact_area` (via `Cell`)
- **Whole sphere**: `whole_sphere()` (six face cells)
- **`Region`**: `cell_union_bound` delegates to the bounding cap’s `cell_union_bound`, matching Go’s behavior
- **`CellID`**: `lsb_for_level` is `pub(crate)` so `expand_at_level` can use the same LSB logic as the rest of the crate

Internal helpers: `lower_bound_cellids`, `are_siblings`, recursive `difference` core.

## Tests

Expanded unit tests cover normalization, containment/intersection probes, merge/intersection/difference, `from_range` edge cases, denormalize, empty union behavior, invalid/duplicate IDs, sibling detection, `whole_sphere`, leaf-cell counts (including five vs six faces and `expand_at_level(0)`), overlapping intersection, and area helpers.

## Out of scope (documented)

- Binary **encode/decode** for `CellUnion` (present in Go; not ported here; see module comment)
- C++-specific surface (protobuf packing, `Iterator`, `Clear`/`Release`, etc.) — Rust uses `Vec<CellID>` idioms instead
- Full **randomized / stress** suites from upstream test files (replaced with targeted tests)

Closes #76

### CODE REVIEW

File: src/s2/cellunion.rs [hunk 1]
```diff
@@ -185,6 +200,190 @@ impl CellUnion {
...
+            }
+        }
+        let mut out = CellUnion(cu);
+        out.normalize();
+        out
+    }
+
+    /// Intersection of a cell union with a single cell ID.
+    pub fn intersection_with_cell_id(x: &Self, id: CellID) -> Self {
+        if x.contains_cellid(&id) {
```
Comment (brawer) [inline src/s2/cellunion.rs:312]: Add a debug assertion, similar to what’s in C++ [here](https://github.com/google/s2geometry/blob/fde443b85c018d2991b7459630932add72ef6f76/src/s2/s2cell_union.cc#L356):

```suggestion
        debug_assert!(id.is_valid(), "invalid id: {:?}", id);
        if x.contains_cellid(&id) {
```

File: src/s2/cellunion.rs [hunk 2]
```diff
@@ -185,6 +200,190 @@ impl CellUnion {
...
+        let id_max = id.range_max();
+        let start = lower_bound_cellids(&x.0, 0, x.0.len(), id.range_min());
+        let mut cu = Vec::new();
+        let mut k = start;
+        while k < x.0.len() && x.0[k] <= id_max {
+            cu.push(x.0[k]);
+            k += 1;
+        }
+        let mut out = CellUnion(cu);
+        out.normalize();
```
Comment (brawer) [inline src/s2/cellunion.rs]: The C++ codebase avoids this (potentially expensive) re-normalization. Instead, it has a debug-only assertion to ensure that the output is normalized in case the input (ie., &self) was: https://github.com/google/s2geometry/blob/fde443b85c018d2991b7459630932add72ef6f76/src/s2/s2cell_union.cc#L367

File: src/s2/cellunion.rs [hunk 3]
```diff
@@ -185,6 +200,190 @@ impl CellUnion {
...
+        out
+    }
+
+    /// Set difference `x - y` for normalized inputs.
+    pub fn difference(x: &Self, y: &Self) -> Self {
+        let mut cu = Vec::new();
+        for xid in &x.0 {
+            cell_union_difference_internal(&mut cu, *xid, y);
+        }
+        CellUnion(cu)
```
Comment (brawer) [inline src/s2/cellunion.rs]: Likewise, add a debug assertion similar to what’s in the C++ codebase [here](https://github.com/google/s2geometry/blob/fde443b85c018d2991b7459630932add72ef6f76/src/s2/s2cell_union.cc#L450).

File: src/s2/cellunion.rs [hunk 4]
```diff
@@ -185,6 +200,190 @@ impl CellUnion {
...
+            k += 1;
+        }
+        let mut out = CellUnion(cu);
+        out.normalize();
+        out
+    }
+
+    /// Set difference `x - y` for normalized inputs.
+    pub fn difference(x: &Self, y: &Self) -> Self {
+        let mut cu = Vec::new();
```
Comment (brawer) [inline src/s2/cellunion.rs:338]: Add a similar comment to what’s in the C++ codebase [here](https://github.com/google/s2geometry/blob/fde443b85c018d2991b7459630932add72ef6f76/src/s2/s2cell_union.cc#L450):

```suggestion
        // TODO: This is approximately O(N*log(N)), but could probably
        // use similar techniques as intersection() to be more efficient.
        let mut cu = Vec::new();
```

File: src/s2/cellunion.rs [hunk 5]
```diff
@@ -242,11 +441,50 @@ impl Region for CellUnion {
...
+}
+
+fn cell_union_difference_internal(out: &mut Vec<CellID>, id: CellID, other: &CellUnion) {
+    if !other.intersects_cellid(&id) {
+        out.push(id);
+        return;
+    }
+    if !other.contains_cellid(&id) {
+        for child in &id.children() {
+            cell_union_difference_internal(out, *child, other);
```
Comment (brawer) [inline src/s2/cellunion.rs]: The C++ version only does this for the first two children; see [here](https://github.com/google/s2geometry/blob/fde443b85c018d2991b7459630932add72ef6f76/src/s2/s2cell_union.cc#L435). Although I must admit that it's not immediately clear to me why this is safe to do.

File: src/s2/cellunion.rs [hunk 6]
```diff
@@ -242,11 +441,50 @@ impl Region for CellUnion {
...
+    mask = !(mask + (mask << 1));
+    let id_masked = d.0 & mask;
+    (a.0 & mask) == id_masked
+        && (b.0 & mask) == id_masked
+        && (c.0 & mask) == id_masked
+        && !d.is_face()
+}
+
+fn cell_union_difference_internal(out: &mut Vec<CellID>, id: CellID, other: &CellUnion) {
+    if !other.intersects_cellid(&id) {
```
Comment (brawer) [inline src/s2/cellunion.rs:507]: The C++ version has a comment [here](https://github.com/google/s2geometry/blob/fde443b85c018d2991b7459630932add72ef6f76/src/s2/s2cell_union.cc#L427):
 
```suggestion
    // Add the difference between `id` and `other` to `out`.
    // If they intersect but the difference is non-empty, divide and conquer.
    if !other.intersects_cellid(&id) {
```

File: src/s2/cellunion.rs [hunk 7]
```diff
@@ -242,11 +441,50 @@ impl Region for CellUnion {
...
+    for i in begin..end {
+        if cells[i] >= id {
+            return i;
+        }
+    }
+    end
+}
+
+fn are_siblings(a: CellID, b: CellID, c: CellID, d: CellID) -> bool {
+    if a.0 ^ b.0 ^ c.0 != d.0 {
```
Comment (brawer) [inline src/s2/cellunion.rs:486]: Add a comment, taken from the C++ codebase [here](https://github.com/google/s2geometry/blob/fde443b85c018d2991b7459630932add72ef6f76/src/s2/s2cell_union.cc#L128-L129):

```suggestion
    // A necessary (but not sufficient) condition is that the XOR of the
    // four cells must be zero.  This is also very fast to test.
    if a.0 ^ b.0 ^ c.0 != d.0 {
```

File: src/s2/cellunion.rs [hunk 8]
```diff
@@ -242,11 +441,50 @@ impl Region for CellUnion {
...
+        return false;
+    }
+    let mut mask = d.lsb() << 1;
```
Comment (brawer) [inline src/s2/cellunion.rs:494]: Add a comment, taken from the C++ codebase [here](https://github.com/google/s2geometry/blob/fde443b85c018d2991b7459630932add72ef6f76/src/s2/s2cell_union.cc#L132):

```suggestion

    // Now we do a slightly more expensive but exact test.  First, compute a
    // mask that blocks out the two bits that encode the child position of
    // `id` with respect to its parent, then check that the other three
    // children all agree with `mask`.
    let mut mask = d.lsb() << 1;
```

File: src/s2/cellunion.rs [hunk 9]
```diff
@@ -185,6 +200,190 @@ impl CellUnion {
...
+        cu
+    }
+
+    /// Normalized union of two cell unions.
+    pub fn union(a: &Self, b: &Self) -> Self {
+        Self::merge(&[a.clone(), b.clone()])
+    }
+
+    /// Intersection of two normalized cell unions.
+    pub fn intersection(a: &Self, b: &Self) -> Self {
```
Comment (brawer) [inline src/s2/cellunion.rs:258]: Add debug assertions similar to those in C++ [here](https://github.com/google/s2geometry/blob/fde443b85c018d2991b7459630932add72ef6f76/src/s2/s2cell_union.cc#L384), especially for the inputs being sorted.

File: src/s2/cellunion.rs [hunk 10]
```diff
@@ -185,6 +200,190 @@ impl CellUnion {
...
+            } else if x[i] < y[j] {
+                cu.push(x[i]);
+                i += 1;
+            } else {
+                cu.push(y[j]);
+                j += 1;
+            }
+        }
+        let mut out = CellUnion(cu);
+        out.normalize();
```
Comment (brawer) [inline src/s2/cellunion.rs]: This looks expensive; compare to C++ [here](https://github.com/google/s2geometry/blob/fde443b85c018d2991b7459630932add72ef6f76/src/s2/s2cell_union.cc#L420) which just has a debug assertion that the output actually is generated in sorted order.

**Review by brawer — COMMENTED** [review]
Very cool! Some nitpicks from comparing your implementation to C++.

**Comment by Louisvranderick** [comment]
Thanks for the detailed review, it was really helpful. My latest commit should cover most of your points and im happy to adjust anything that still doesn’t look right.

**Review by brawer — APPROVED** [review]
Thank you!

File: src/s2/cellunion.rs (hunk 10 above)
Comment (brawer) [reply src/s2/cellunion.rs]: OK, let’s merge as-is.
